### ISSUE TITLE
It costs 6+s time to run post processing in your demo?

### ISSUE THREAD

**Issue by Mengmengbai** [Atlas200dk/sample-imageinpainting-HiFill#4]
I test the post runtime beyound 6s.
But you paper pay more attention to talk the inpainting model runtime?

8  Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz

**Comment by duxingren14** [comment]
Thanks for your interests in our repo. The runtime speed mentioned in the paper includes both model inference, plus the pre- and post-processing. The reason why this version is slower, is it uses only CPU for pre- and post-processing. Especially the matmul operation is perform using numpy 
 (np.matmul) rather than tensorflow (tf.matmul). Also, resize_ave can be implemented using tf.

This version is to maintain the consistency of the graph in pb file and taht in huawei's omg file.

**Comment by Mengmengbai** [comment]
I implement your post processing with GPU, GTX 1080Ti. However, when I test your model with 512*512 input size images, it costs 618.03ms. The model takes 32ms to infer, the post processing spends 586.03ms. I think it is unfair to compare your part solution running time whit the other total solution in your paper. Moreover, for the 3072*3072 size images, I think your post processing will cost huge time, because there is a huge matmul operation in your post processing.

**Comment by Mengmengbai** [comment]
Matrix multiplication operation, (1,32,32,1024) tensor matmul (1,32,1024,12288), takes 177.146ms.
Tensorflow 1.13.1, CUDA 10.1 cuDNN 7.6, CPU i7-7700K, input size 512*512

**Comment by duxingren14** [comment]
Thanks for the feedback. In the provided implementation, we generally resize images to 3K for conveinece. Actually, we have more sophisticted implementations, that is resize an image to multiples of 512. For example, if the input image is hxw, we resize it to h'=(h//512)*512, w'=(w//512)*512.  This will save much more time, when image size is below 3K. The test results in our paper are based on this implementation.  

But still, I think the inference time reported by you is a more than we tested. I have no idea what is the cause. We will try to release our codes in the future.

**Comment by Mengmengbai** [comment]
I have tested the inference time of your published model with input 512*512 size image. The result is the same to your paper said. It costs 31ms around, but even the post processing operation takes  about 580ms. I don't know how you implement the matmul operation. I doubt that your post-processing is completed in 1ms. I cannot believe it. Matrix multiplication operation, (1,32,32,1024) tensor matmul (1,32,1024,12288), can be completed in 1ms, with GTX 1080Ti. Even that, how about the 3072*3072 input size? the matrix multiplication operation is more huge than 512*512 input size.

**Comment by duxingren14** [comment]
When the image sz <= 512, we did not need post processing.
Just wanna confirm that  , the input of matmul is (1,1,1024,1024)  (1,1,1024,12288)?

**Comment by Mengmengbai** [comment]
If the image size < 512, you don't use the post processing? But the result is so blurry.
the input is (1,32,32,1024) (1,32,1024,12288). GPU 1080Ti will take 177ms to complement it.
How about 2048*2048 size image?
the input matmul will be (1,32,32,1024) (1,32,1024,27648). I think it costs more time than 512 size image to run on GTX 1080ti
If you use post processing with 2048 image size, I am very curious about how you quickly, just few milliseconds, calculate such 
a large matrix operation with GTX 1080ti.
If that, I think this is a very critical point what does not appear in your paper.

**Comment by duxingren14** [comment]
Oh, I am just trying to help. As the time claimed in the paper definitely covers the time for post-processing. 
I wish I could help you more to figure out the issues.

In addition, I should have release the complete set o source codes in the near future, though it will take time to get company's approval. Once you get the chance to read the official codes, you will know more.

when img sz <=  512, the output of the refinement network is O. then the final result is the masked region of O merged with the unmasked region of input image I. In that case, you will not see the blurry issue generally. Here when I say we do not need post-processing, I mean the residual aggregation module.

**Comment by Mengmengbai** [comment]
Sincerely, you didn't convince me.
In theory, the huge matmul can not be complemented with few milliseconds on GTX 1080Ti.

Even your 512*512 input size is the same to your paper without post processing, how do you explain your 1024*1024 and 2048*2048 image?
I have tested 1024*1024 and 2048*2048 running time, and they both are the same to your paper.
Doesn't your post processing take time?

Moreover, the runing time of contextual attention layer is related to the input size. 
In other words, the larger input size, the more it time costs. 
Your post processing takes 580ms around to run 512*512 size image. 
Are 1024*1024 or 2048*2048 images faster than 512*512?

**Comment by duxingren14** [comment]
if img size is 512x512, networking 25ms 
if img size is 1024,  networking 25ms + post-prcessing (6ms),  matmul inputs size: [1,1,1024,1024],  [1,1,1024, 1024*3]
if img size is 2048, networking 25ms + post-processing (12ms) matmul inputs size: [1,1,1024,1024],  [1,1,1024, 4096*3]
if img size is 4096, networking 25ms + post-processing (62ms), matmul input size: [1, 1, 1024, 1024], [1,1,,1024, 4096*4*3]

the matmul in tensorflow can also be implemented using conv_transpose, which is faster as less reshaping and transpose are needed. Probably it is a bit too complicated to explain here. But just wish you to be patient enough to wait for the complete sets of source codes.

**Comment by Mengmengbai** [comment]
I know how to implement it with conv_transpose, I will test it again.
One more thing, I test a lot inpainting networks, PC is the best one.
https://www.nvidia.com/research/inpainting/selection
this is the PC online test site, the result is far from your paper about PC results

**Comment by renleidewenming** [comment]
In the generator, the input 512x512 image is first downsampled to 256x256 and the inpainting result is 256x256 too, which is then upsampled to 512x512 as the final output of the generator. So in the test time, if image size is 512x512, the final result is actually upsampled from 256x256, which I think will be blurry. Am I right? I think postprocess is also necessary for 512x512 image.

**Comment by duxingren14** [comment]
Please find the tensorflow implementation of the post-processing module from here
https://github.com/duxingren14/Hifill-tensorflow
